src/Compilers/CSharp/Test/Symbol/Symbols/ModuleInitializersTests.Accessibility.cs

```diff
@@ -0,0 +1,265 @@
+
+class C
+{
+    [ModuleInitializer]
+    " + keywords + @" static void M() => Console.WriteLine(""C.M"");
+}
+
+class Program 
+{
+    static void Main() => Console.WriteLine(""Program.Main"");
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+            CompileAndVerify(source, parseOptions: s_parseOptions, expectedOutput: @"
+C.M
+Program.Main");
+        }
+
+        [Theory]
+        [InlineData("public")]
+        [InlineData("internal")]
+        public void AllowedTopLevelTypeAccessibility(string keywords)
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+" + keywords + @" class C
+{
+    [ModuleInitializer]
+    public static void M() => Console.WriteLine(""C.M"");
+}
+
+class Program 
+{
+    static void Main() => Console.WriteLine(""Program.Main"");
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+            CompileAndVerify(source, parseOptions: s_parseOptions, expectedOutput: @"
+C.M
+Program.Main");
+        }
+
+        [Theory]
+        [InlineData("private")]
+        [InlineData("protected")]
+        [InlineData("private protected")]
+        public void DisallowedNestedTypeAccessibility(string keywords)
+        {
+            string source = @"
+using System.Runtime.CompilerServices;
+
+public class C
+{
+    " + keywords + @" class Nested
+    {
+        [ModuleInitializer]
+        public static void M() { }
+    }
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+            var compilation = CreateCompilation(source, parseOptions: s_parseOptions);
+            compilation.VerifyEmitDiagnostics(
+                // (8,10): error CS8794: Module initializer method 'M' must be accessible outside top-level type 'C'
+                //         [ModuleInitializer]
+                Diagnostic(ErrorCode.ERR_ModuleInitializerMethodMustBeAccessibleOutsideTopLevelType, "ModuleInitializer").WithArguments("M", "C").WithLocation(8, 10)
+                );
+        }
+
+        [Theory]
+        [InlineData("public")]
+        [InlineData("internal")]
+        [InlineData("protected internal")]
+        public void AllowedNestedTypeAccessibility(string keywords)
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+public class C
+{
+    " + keywords + @" class Nested
+    {
+        [ModuleInitializer]
+        public static void M() => Console.WriteLine(""C.M"");
+    }
+}
+
+class Program 
+{
+    static void Main() => Console.WriteLine(""Program.Main"");
+}
+
+namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : System.Attribute { } }
+";
+            CompileAndVerify(source, parseOptions: s_parseOptions, expectedOutput: @"
+C.M
+Program.Main");
+        }
+
+        [Fact]
+        public void MayBeDeclaredByStruct()
```

Comment:
>MayBeDeclaredByStruct [](start = 20, length = 21)

This is a good test. However, it is not clear to me why it is in the accessibility bucket.

Reply:
It doesn't look like we are testing various accessibilities. Was the intent to actually test accessibility? 

---
In reply to: [420264798](https://github.com/dotnet/roslyn/pull/43301#discussion_r420264798) [](ancestors = 420264798)

Reply:
It probably shouldn't be in this bucket. Do you see value in repeating the class accessibility tests for structs and interfaces, or should I move these to the main file?

Reply:
> Do you see value in repeating the class accessibility tests for structs and interfaces, or should I move these to the main file?

I do not think accessibility check depends on the kind of types involved. We could have accessibility tests like that for completeness, but I don't think they are strictly required. Perhaps they can be folder into the theory tests above in some way.



---
In reply to: [420522461](https://github.com/dotnet/roslyn/pull/43301#discussion_r420522461) [](ancestors = 420522461)